File: backend/app/actor/ali/linh/response/Output.scala
package actor.ali.response

import actor.ali.search.model.Question

case class Output(content: Seq[OutputItem], suggestions: Option[Suggestions] = None)

case class OutputItem(kind: OutputKind,
                      text: Option[String])



case class Suggestions(label: String, items: Seq[String])



case object Output {
    def single(o: OutputItem):Output = Output(Seq(o))
    def single(o: OutputItem, s: Suggestions):Output = Output(Seq(o), Some(s))

    def answersOf(q: Question, suggestions: Option[Suggestions] = None):Output = {
        val answers = q.answers.map(OutputItem.text)
        Output(answers, suggestions)
    }
}

case object OutputItem {
    def text(text: String):OutputItem = OutputItem(OutputKind.PLAIN, Some(text))

    def logg(text: String):OutputItem = OutputItem(OutputKind.LOG, Some(text))

    def success(text: String):OutputItem = OutputItem(OutputKind.SUCCESS, Some(text))

    def error(text: String):OutputItem = OutputItem(OutputKind.ERROR, Some(text))
}

File: frontend/src/app/models/output.ts
import {Content, ContentKind, OutputKind} from './config';
import {OutputItem} from './outputItem';
import {Suggestions} from './suggestions';

export class Output implements Content{
  kind: ContentKind = ContentKind.OUT;
  content: OutputItem[];
  suggestions ?: Suggestions;

  constructor(content: OutputItem[], suggestions?: Suggestions) {
    this.content = content;
    this.suggestions = suggestions;
  }

  static plain(msg: string):Output {
    const item = new OutputItem(OutputKind.PLAIN, msg)
    return new Output([item])
  }

  static logg(msg: string):Output {
    const item = new OutputItem(OutputKind.LOG, msg)
    return new Output([item])
  }

  static success(msg: string):Output {
    const item = new OutputItem(OutputKind.SUCCESS, msg)
    return new Output([item])
  }

  static error(msg: string):Output {
    const item = new OutputItem(OutputKind.ERROR, msg)
    return new Output([item])
  }

  static fromJson(parsedJson: any):Output {
    const o = Object.assign(Object.create(Output.prototype), parsedJson);
    o.kind = ContentKind.OUT; //todo backend doesn't assign this atm
    return o;
  }
}
